### ISSUE TITLE
Scatter destination check always inserts a relayout copy

### ISSUE THREAD

**Issue by tdoublep** [torch-spyre/torch-spyre#4115]
## Problem

`_enforce_scatter_destination_layout` never actually decides whether a scatter destination is compliant. It always inserts a relayout copy, even when the destination layout is already exactly what the pass wants.

## Why

The check computes the write coordinates with `indirect_sizes=None` (`enforce_indirect_access_layout.py`):

```python
write_coords = device_coordinates(target_stl, write_dep, None)
```

When `indirect_sizes` is `None`, `compute_coordinates` skips indirect symbols entirely (`views.py`, the `else: continue` branch). So no coordinate ever carries an `IndirectAccess` marker, and the loop right below — which searches those coordinates for exactly that marker — always comes up empty.

`is_compliant` starts out `False` and is only ever set inside `if indirect_stride_idxs:`. So "could not determine" silently becomes "not compliant", and the copy is inserted unconditionally. The `compliant=` debug line lives in the same skipped branch, so the diagnostic disappears exactly when you would want it.

## Reproducer

The destination's indexed dim is at device position 0 — what the pass requires — so no copy should be inserted.

```python
import torch
import torch_spyre  # noqa: F401

torch_spyre._autoload()
torch.spyre.set_device(0)
# A layout-carrying .to() skips _lazy_init, so start the runtime first.
torch.zeros(1, dtype=torch.float16).to("spyre")

from torch_spyre._C import SpyreTensorLayout, get_device_dtype, get_elem_in_stick

SLOTS, HEADS, HD, T = 8, 2, 64, 4
eps = get_elem_in_stick(torch.float16)
sticks = HD // eps

layout = SpyreTensorLayout(
    device_size=[SLOTS, HEADS, sticks, eps],
    stride_map=[HEADS * sticks * eps, sticks * eps, eps, 1],
    device_dtype=get_device_dtype(torch.float16),
)

dst = torch.zeros(SLOTS, HEADS, HD, dtype=torch.float16).to("spyre", device_layout=layout)
src = torch.randn(T, HEADS, HD, dtype=torch.float16).to("spyre")
idx = torch.arange(T, dtype=torch.int32).to("spyre")


def scatter(dst, idx, src):
    dst.index_copy_(0, idx, src)


torch.compile(scatter, dynamic=False)(dst, idx, src)
print("max abs diff:", (dst.cpu()[:T].float() - src.cpu().float()).abs().max().item())
```

Run with `SPYRE_INDUCTOR_LOG=1 SPYRE_INDUCTOR_LOG_LEVEL=DEBUG`. Observed:

```
scatter_destination_check: inserting mutation relayout copy for buf0
```

and no `compliant=` line anywhere in the log.

## Impact

Results stay correct, so this is pure overhead. Going through vLLM on granite-3.3-8b, it adds 4 restickify ops per layer over the KV cache (1.31 MB read + 1.31 MB write each), because the cache is already handed a slot-major destination layout. Removing the copies improves end-to-end decode latency by roughly 2% in `vllm bench latency` (batch 1, input 64 / output 64, 10 iters).

## Fix

Build the real subs and sizes with `_scatter_access_subs_and_sizes` — the same helper `_insert_mutation_relayout_copy` already uses — and fall back to the copy if a size cannot be resolved.

Patch (on top of `e6aa948`): https://github.com/tdoublep/torch-spyre/tree/fix/scatter-destination-check-indirect-sizes

With that applied, the reproducer above logs `indirect_device_pos=[0], expected=[0], compliant=True` and inserts no copy. `max abs diff` is `0.0` either way.

**Comment by avery-blanchard** [comment]
@tdoublep I opened a PR dropping this copy, let me know if you still hit issues with this?

**Comment by tdoublep** [comment]
Thanks @avery-blanchard - I can confirm that this is giving me an 8% reduction in E2E latency when running following benchmark:
```
SPYRE_NUM_CPUS=8 uv run --no-sync vllm bench latency \
    --model ibm-granite/granite-3.3-8b-instruct \
    --input-len 64 --output-len 512 --batch-size 1 \
    --num-iters-warmup 2 --num-iters 3 --max-model-len 576 \
    -cc.compile_sizes='[1,64]'
```
